#163584006 Implement tags on Room CRUD operations

bevkololi opened the pull request:
#### What does this PR do?
Allows admins to include tags in room CRUD operations i.e Creating rooms and Updating rooms.

#### Description of Task to be completed?
Admin's adding and editing rooms should select tags that are to be associated with the room. Deleting rooms should remove the relationship between the room and tags.

#### How should this be manually tested?
* Clone the branch and follow the following steps to [setup](https://github.com/andela/mrm_api).
* git checkout to `ft-implement-tags-in-rooms-163584006` and run the following mutations:

Create room:
```
mutation {
        createRoom(
            name: "Room 2", roomType: "Meeting", capacity: 4, floorId: 2, officeId: 1, roomTags: [1, 3],
      calendarId:"andela.com_3836323338323230343935@resource.calendar.google.com",
            imageUrl: "https://www.officelovin.com/wp-content/uploads/2016/10/andela-office-main-1.jpg") {  
            room {
                id
                name
                roomType
                capacity
                floorId
                imageUrl
                roomTags {
                  name
                  description
                }
            }
        }
    }
```
Update room:
```
mutation{
    updateRoom(roomId: 2, name: "Jinja", capacity: 8,
    roomType: "board room", roomTags: [1, 3]){
        room{
            name
            capacity
            roomType
          	roomTags {
              name
              color
              description
            }
        }
    }
}
```

#### Any background context you want to provide?
An admin can add/update one or more tags in a room. Deleting a room automatically deletes the relationship between the room and its tags.

#### Screenshots
* On creating a room:
<img width="1440" alt="screenshot 2019-01-31 at 18 50 38" src="https://user-images.githubusercontent.com/26184534/52068144-230a3a00-258d-11e9-8d1d-177cfd70f705.png">

* On creating a room with a no-existing tag
<img width="1440" alt="screenshot 2019-02-04 at 12 19 42" src="https://user-images.githubusercontent.com/26184534/52199498-51cb2d80-2877-11e9-89f8-014c18a2f3af.png">

*On updating a room
<img width="1440" alt="screenshot 2019-01-31 at 19 22 25" src="https://user-images.githubusercontent.com/26184534/52068349-93b15680-258d-11e9-9d8e-6d1a68a15fd4.png">

*On getting all rooms
<img width="1440" alt="screenshot 2019-02-04 at 20 07 23" src="https://user-images.githubusercontent.com/26184534/52224768-dfc80800-28b9-11e9-8ebf-7620b99e9cd3.png">


#### What are the relevant pivotal tracker stories?
[#163584006](https://www.pivotaltracker.com/story/show/163584006)

caveinn commented on api/room/schema.py:
```diff
@@ -79,8 +91,14 @@ def mutate(self, info, office_id, **kwargs):
             kwargs.pop('block_id')
+        # pop the room_tags kwargs to create the room object
+        # this is because it populates the association table.
+        room_tags = ''
```
Hello @bevkololi, I think for the room_tag you should assign it to an empty list so that it id intuitive that the variable expects to be populated with a list of items.

bevkololi replied:
Thank you for that @caveinn, working on it.

bevkololi replied:
This has been updated @caveinn

RonKbS commented on api/room/schema.py:
```diff
@@ -114,22 +132,27 @@ class Arguments:
+            room_tags = kwargs.pop('room_tags')
+        update_entity_fields(room, **kwargs)
+        new_tags = room.room_tags
+        new_tags.clear()
```
Kindly share the purpose of the .clear() method

bevkololi replied:
@RonKbS, the relationship table room_tags comes as a list, so when I'm editing the tags in a room, say a room has tags [1, 3, 4] and I want to update the tags to [1, 3], I chose to clear the list using the python method .clear() and append the new tags.

RonKbS commented on api/room/schema.py:
```diff
@@ -20,6 +21,16 @@ class Meta:
+    for tag in room_tags:
+        room_tag = TagModel.query.filter_by(id=tag).first()
+        if not room_tag:
+            room.delete()
```
Please share the reason for the .delete() because the method name implies that saving is being carried out.

bevkololi replied:
@RonKbS, this is because whether the tag id is valid or not, the room is still saved (kindly test this locally). I therefore have to remove it if the mutation has an invalid tag.

cwizard2011 replied:
@bevkololi please respond to @RonKbS

RonKbS replied:
I see you've pushed some changes and removed the room.save() on line 102 just before this function is called. The .delete() method should be unnecessary now.

bevkololi replied:
It still saves the room @RonKbS. That line did not affect it.

bevkololi replied:
Worked on this. Thanks.

kampkelly commented on api/room/schema.py:
```diff
@@ -20,6 +21,16 @@ class Meta:
+            raise GraphQLError("Tag not found")
+        room.room_tags.append(room_tag)
+        room.save()
+
```
Good work @bevkololi. But I am of the opinion that there can be a better way to refactor this to avoid calling save multiple times?

bevkololi replied:
Yes, thanks @kampkelly, as I had already saved it in the save_room_tags() method i do not need to save it again. Updating this as we speak.

RonKbS commented on api/room/schema.py:
```diff
@@ -114,22 +130,27 @@ class Arguments:
+            room_tags = kwargs.pop('room_tags')
+        update_entity_fields(room, **kwargs)
+        previous_tags = room.room_tags
+        previous_tags.clear()
```
I've tried out `room.room_tags.clear()` and it seems to work as well.
Is there a reason for using a pointer to call .clear()?

kampkelly replied:
I guess she is doing that to make it descriptive by indicating that it is the previous tags that are being cleared.

bevkololi replied:
Was trying to make it more readable but sure this can be worked on too.

dnuwa commented on fixtures/room/create_room_fixtures.py:
```diff
@@ -25,17 +31,42 @@
+non_existing_tag_mutation = '''
+    mutation {
+        createRoom(
+            name: "Mbarara", roomType: "Meeting", capacity: 4, floorId: 4, officeId: 1,,
```
@bevkololi good work, though I'm wondering why you added an extra comma here and on line 6.

bevkololi replied:
Thanks @dnuwa, let me fix this.

Bonifase commented on api/room/schema.py:
```diff
@@ -20,6 +21,15 @@ class Meta:
+def save_room_tags(room, room_tags):
+    for tag in room_tags:
+        room_tag = TagModel.query.filter_by(id=tag).first()
+        if not room_tag:
```
Since there are more than one tag coming from the user, would it be appropriate to be more specific in the error message?

bevkololi replied:
I have updated the code to state all the tags that are invalid @Bonifase. Thanks for pointing that out.

Bonifase commented on api/room/schema.py:
```diff
@@ -79,8 +90,13 @@ def mutate(self, info, office_id, **kwargs):
         # remove block ID from kwargs, can't be saved in rooms model
         if kwargs.get('block_id'):
             kwargs.pop('block_id')
+        # pop the room_tags kwargs to create the room object
```
Remove the commented out codes from this file

Bonifase commented on tests/test_rooms/test_create_room.py:
```diff
@@ -101,3 +102,13 @@ def test_room_creation_with_invalid_calendar_id(self):
+        """
+        CommonTestCases.admin_token_assert_in(
+            self,
+            non_existing_tag_mutation,
```
Since we know all the tags that are supposed to be there by name, use tag names in your error messages and avoid using integers to add more clarity to users.

bevkololi replied:
@Bonifase, your concerns are noted. However, this might nit be applicable when a tag does not exist completely, the tag does not have a name, only way to raise an error is to use the integer the user had input.

bevkololi replied:
This has been updated to "Tag id" for more clarity @Bonifase, since it is impossible to use tag name for non-existent tags.

bevkololi commented:
> @bevkololi great work so far but looks like you implemented tags in only mutations. I believe you were meant to do all CRUD.

@joshuaocero, I have implemented them all in this PR. I 

> @bevkololi great work so far but looks like you implemented tags in only mutations. I believe you were meant to do all CRUD.

@joshuaocero, I had implemented all all crud operations. I have updated the screenshots to include querying for rooms.

bevkololi commented:
> @bevkololi great work so far but looks like you implemented tags in only mutations. I believe you were meant to do all CRUD.

 I had implemented all all crud operations. I have updated the screenshots to include querying for rooms.

bevkololi commented:
> Kindly check on your commit messages. Please use the guide that I shared on the slack channel last weeks

I have updated this @Bonifase.
